### PR TITLE
[SPARK-21848][SQL] Add trait UserDefinedExpression to identify user-defined functions

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?

Add trait UserDefinedExpression to identify user-defined functions.
UDF can be expensive. In optimizer we may need to avoid executing UDF multiple times.
E.g.
```scala
table.select(UDF as 'a).select('a, ('a + 1) as 'b)
```
If UDF is expensive in this case, optimizer should not collapse the project to
```scala
table.select(UDF as 'a, (UDF+1) as 'b)
```

Currently UDF classes like PythonUDF, HiveGenericUDF are not defined in catalyst. 
This PR is to add a new trait to make it easier to identify user-defined functions.

## How was this patch tested?

Unit test

### CODE REVIEW

**Review by srowen — CHANGES_REQUESTED** [review]
Why a trait? It doesn't seem to be used other than to mark.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala [hunk 1]
```diff
@@ -22,6 +22,9 @@ import org.apache.spark.sql.catalyst.{CatalystTypeConverters, InternalRow}
 import org.apache.spark.sql.catalyst.expressions.codegen._
 import org.apache.spark.sql.types.DataType
 
+// Trait for identifying user-defined functions.
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala]:
```Scala
/**
 * Common base trait for user-defined functions, including ScalaUDF, ScalaUDAF, PythonUDF, 
 * HiveSimpleUDF, HiveGenericUDF, and HiveUDAFFunction.
 */
```

**Comment by gatorsmile** [comment]
Yes. This is a potential optimization we should do, but it should be cost based. Let us improve the resolution logics using this common base trait in this PR first.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala [hunk 2]
```diff
@@ -23,6 +23,12 @@ import org.apache.spark.sql.catalyst.expressions.codegen._
 import org.apache.spark.sql.types.DataType
 
 /**
+ * Common base trait for user-defined functions, including ScalaUDF, ScalaUDAF, PythonUDF,
+ * HiveSimpleUDF, HiveGenericUDF, and HiveUDAFFunction.
+ */
+trait UDFType
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala]: maybe UserDefinedExpression?

Comment (rxin) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala]: also sort of weird to put this in ScalaUDF.scala. Either put it in Expression.scala, or create its own file ...

**Comment by jiangxb1987** [comment]
I'm +1 on this, currently we are pattern matching `ScalaUDF` in optimize rules which is suboptional.

**Comment by jiangxb1987** [comment]
I'm +1 on this, it should be useful to pattern match the trait instead of handling ScalaUDF/ScalaUDAF/PythonUDF... in future optimize rules.

**Comment by viirya** [comment]
It is better to review PR title and description to reflect the change from `UDFType` to `UserDefinedExpression`.

**Comment by viirya** [comment]
retest this please.

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

**Pull request closed without merging**